### PR TITLE
ArC: build profile enablement for additional beans

### PR DESCRIPTION
**Pull request by mkouba**
- i.e. support IfBuildProfile, UnlessBuildProfile, IfBuildProperty and UnlessBuildProperty annotations declared on additional beans classes

Zulip discussion: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Custom.20.20Extension.3A.20Ensuring.20Test.20Mode.20and.20Configuring

### CODE REVIEW

File: extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/AdditionalBeanWithIfBuildProfileTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
+import io.quarkus.builder.BuildContext;
+import io.quarkus.builder.BuildStep;
+import io.quarkus.test.QuarkusUnitTest;
+
+@Disabled("For some reason the indexing of AdditionalIfBuildProfileBean fails - the class resource cannot be found by the CL")
```
Comment (mkouba) [inline extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/AdditionalBeanWithIfBuildProfileTest.java:17]: This is unfortunate but I have no idea how to fix it now.

**Review by radcortez — APPROVED** [review]

**Review by geoand — APPROVED** [review]

**Comment by mkouba** [comment]
Unfortunately, this change causes a cycle in the build chain: 
```
Caused by: java.lang.RuntimeException: io.quarkus.builder.ChainBuildException: Cycle detected:
		   io.quarkus.arc.deployment.BeanArchiveProcessor#build produced class io.quarkus.arc.deployment.BeanArchiveIndexBuildItem
		to io.quarkus.arc.deployment.BuildTimeEnabledProcessor#ifBuildProfile produced class io.quarkus.arc.deployment.BuildTimeConditionBuildItem
		to io.quarkus.resteasy.reactive.common.deployment.ResteasyReactiveCommonProcessor#handleApplication produced class io.quarkus.resteasy.reactive.common.deployment.ApplicationResultBuildItem
		to io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveScanningProcessor#scanForContextResolvers produced class io.quarkus.arc.deployment.AdditionalBeanBuildItem
		to io.quarkus.arc.deployment.BeanArchiveProcessor#build
```

@geoand Do you have an idea how to break this cycle?

**Comment by geoand** [comment]
Hm... I am sure we will be able to...

**Comment by mkouba** [comment]
> Hm... I am sure we will be able to...

@geoand would you care to work on this? :wink:

**Comment by geoand** [comment]
I am due for PTO, so it won't be soon

**Comment by mkouba** [comment]
> I am due for PTO, so it won't be soon

Ok, np.

**Comment by mkouba** [comment]
> I am due for PTO, so it won't be soon

@geoand back from PTO yet? :D

**Comment by geoand** [comment]
If I said no would you believe me? 😆

**Comment by geoand** [comment]
@mkouba can you check https://github.com/quarkusio/quarkus/pull/35793 please?

**Comment by mkouba** [comment]
> @mkouba can you check #35793 please?

@geoand I have no idea what those changes actually do ;-). Anyway, could you add that commit to this PR so that we can see if it helps...

**Comment by mkouba** [comment]
> Feel free to pull that commit into this PR so we can see in action

_"Maintainers are allowed to edit this pull request."_ :smile: but for sure I'm on it!

**Comment by mkouba** [comment]
```
Caused by: io.quarkus.builder.ChainBuildException: Cycle detected:
		   io.quarkus.arc.deployment.BeanArchiveProcessor#build produced class io.quarkus.arc.deployment.BeanArchiveIndexBuildItem
		to io.quarkus.arc.deployment.BuildTimeEnabledProcessor#unlessBuildProperty produced class io.quarkus.arc.deployment.BuildTimeConditionBuildItem
		to io.quarkus.resteasy.reactive.common.deployment.ResteasyReactiveCommonProcessor#handleApplication produced class io.quarkus.resteasy.reactive.common.deployment.ApplicationResultBuildItem
		to io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveScanningProcessor#scanForDynamicFeatures produced class io.quarkus.resteasy.reactive.spi.DynamicFeatureBuildItem
		to io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveCDIProcessor#additionalBeans produced class io.quarkus.arc.deployment.AdditionalBeanBuildItem
		to io.quarkus.arc.deployment.BeanArchiveProcessor#build
```

@geoand ^ ?

**Comment by geoand** [comment]
Hm... I don't really like where this going so I'll have to rethink it.

**Comment by geoand** [comment]
What I had to do is so so ugly but I don't see any other way around it

**Comment by mkouba** [comment]
> What I had to do is so so ugly but I don't see any other way around it

And it does not seem to help :-(

```
Caused by: io.quarkus.builder.ChainBuildException: Cycle detected:
		   io.quarkus.arc.deployment.BuildTimeEnabledProcessor#unlessBuildProperty produced class io.quarkus.arc.deployment.BuildTimeConditionBuildItem
		to io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveScanningProcessor#scanForContextResolvers produced class io.quarkus.arc.deployment.AdditionalBeanBuildItem
		to io.quarkus.arc.deployment.BeanArchiveProcessor#build produced class io.quarkus.arc.deployment.BeanArchiveIndexBuildItem
		to io.quarkus.arc.deployment.BuildTimeEnabledProcessor#unlessBuildProperty
```

**Comment by mkouba** [comment]
I'm closing this pull request as it seems that we can't easily solve the problem with cycle in the build steps.

**Pull request closed without merging**